src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,137 @@
+using namespace chip::SetupPayload;
+using namespace chip::Encoding::LittleEndian;
+
+static uint8_t * sHashInput;
+
+AdditionalDataPayloadGenerator::AdditionalDataPayloadGenerator(uint16_t lifetimeCounter, char * serialNumberBuffer,
+                                                               size_t serialNumberBufferSize) :
+    mLifetimeCounter(lifetimeCounter),
+    mSerialNumberBuffer(serialNumberBuffer), mSerialNumberBufferSize(serialNumberBufferSize)
+{
+    sHashInput =
+        reinterpret_cast<uint8_t *>(chip::Platform::MemoryAlloc(ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + serialNumberBufferSize));
+}
+
+AdditionalDataPayloadGenerator::~AdditionalDataPayloadGenerator()
+{
+    if (sHashInput != nullptr)
+    {
+        chip::Platform::MemoryFree(sHashInput);
+        sHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle,
+                                                                         bool enableRotatingDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(rotatingDeviceIdBuffer, ROTATING_DEVICE_ID_LENGTH * 2 + 1, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdBufferOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char lifetimeCounterBuf[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    uint8_t outputBuffer[ROTATING_DEVICE_ID_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter hashInputBufferWriter(sHashInput, ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize);
+    BufferWriter outputBuferWriter(&outputBuffer[0], ArraySize(outputBuffer));
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= ROTATING_DEVICE_ID_LENGTH * 2 + 1, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    snprintf(lifetimeCounterBuf, sizeof(lifetimeCounterBuf), "%09u", mLifetimeCounter);
```

Comment:
The question exists again: why not encode the lifetime by just calling `hashInputBufferWriter.Put16(static_cast<uint16_t>(mLifeTimeCounter))` (or similar). Why use ASCII decimal serialization here?

Reply:
my initial thought is to have a 0-padded string of the counter to help with sharding purposes, but i am okay with putting it as a number